### ISSUE TITLE
Elm.Syntax.Pattern.patternRange

### ISSUE THREAD

**Issue by Janiczek** [stil4m/elm-syntax#2]
This might be worth including in the package, similarly to `Elm.Syntax.Exposing.topLevelExposeRange`.

```
patternRange : Elm.Syntax.Pattern.Pattern -> Range
patternRange pattern =
    case pattern of
        Elm.Syntax.Pattern.AllPattern range ->
            range

        Elm.Syntax.Pattern.UnitPattern range ->
            range

        Elm.Syntax.Pattern.CharPattern _ range ->
            range

        Elm.Syntax.Pattern.StringPattern _ range ->
            range

        Elm.Syntax.Pattern.IntPattern _ range ->
            range

        Elm.Syntax.Pattern.FloatPattern _ range ->
            range

        Elm.Syntax.Pattern.TuplePattern _ range ->
            range

        Elm.Syntax.Pattern.RecordPattern _ range ->
            range

        Elm.Syntax.Pattern.UnConsPattern _ _ range ->
            range

        Elm.Syntax.Pattern.ListPattern _ range ->
            range

        Elm.Syntax.Pattern.VarPattern _ range ->
            range

        Elm.Syntax.Pattern.NamedPattern _ _ range ->
            range

        Elm.Syntax.Pattern.QualifiedNamePattern _ range ->
            range

        Elm.Syntax.Pattern.AsPattern _ _ range ->
            range

        Elm.Syntax.Pattern.ParenthesizedPattern _ range ->
            range
```

**Comment by Janiczek** [comment]
Or, refactoring the types to put the range out of the union type! :)

**Comment by Janiczek** [comment]
Similarly, `Elm.Syntax.TypeAnnotation.typeAnnotationRange`:
```
typeAnnotationRange : Elm.Syntax.TypeAnnotation.TypeAnnotation -> Range
typeAnnotationRange typeAnnotation =
    case typeAnnotation of
        Elm.Syntax.TypeAnnotation.GenericType _ range ->
            range

        Elm.Syntax.TypeAnnotation.Typed _ _ _ range ->
            range

        Elm.Syntax.TypeAnnotation.Unit range ->
            range

        Elm.Syntax.TypeAnnotation.Tupled _ range ->
            range

        Elm.Syntax.TypeAnnotation.Record _ range ->
            range

        Elm.Syntax.TypeAnnotation.GenericRecord _ _ range ->
            range

        Elm.Syntax.TypeAnnotation.FunctionTypeAnnotation _ _ range ->
            range
```

**Comment by stil4m** [comment]
I think extracting the range out of the union type is a better idea. It enforces to *have* a range instead of one value constructor to have none. Expressions already have this.

I'm thinking about creating a type alias as:

```
type alias Ranged a = 
    (Range, a)
```

which allows you to use it as a `Ranged Pattern` or `Ranged Expression`. This change would result in a major version bump.
What do you think?

**Comment by Janiczek** [comment]
Seems like a good idea to me!

**Comment by stil4m** [comment]
I've implemented it in the issue-2 branch.

* https://github.com/stil4m/elm-syntax/blob/issue-2/src/Elm/Syntax/TypeAnnotation.elm#L23
* https://github.com/stil4m/elm-syntax/blob/issue-2/src/Elm/Syntax/Expression.elm#L107
* and so on.

Using it would be as following:

* https://github.com/stil4m/elm-syntax/blob/issue-2/src/Elm/Writer.elm#L140
* https://github.com/stil4m/elm-syntax/blob/issue-2/src/Elm/Writer.elm#L335

Feedback is welcome :)

**Comment by Janiczek** [comment]
From looking at the commits in the issue-2 branch, seems good to me! I don't currently have time to incorporate that into my code, so no field report, but if I find anything later, I'll open an issue again :)

**Comment by stil4m** [comment]
Alright. I'll merge it into master and release it together with #3 into a new major version. I've tested it with `elm-test` and will field test it in `elm-analyse`.

At least the structure becomes more consistent, thus that is +1.

**Comment by stil4m** [comment]
Fixed in 3.0.0
